Refactor Test&Troubleshoot (#2666)
astronomer/docs: astro/cli/test-your-astro-project-locally.md

---

> sidebar_label: 'Test your project'

Reviewer: There will be some follow-up work on this page after this PR merges to address the scope of the doc and separate into two pages that address testing DAGs specifically vs testing the entire project since there are different scenarios that we can then expand on (re upgrade runtime) after we make the distinction more clear. 

See https://github.com/astronomer/docs/issues/2847

title: 'Test your Astro project'
id: test-your-astro-project-locally

> description: Check your Astro project for errors before you run it locally or deploy to Astro.

Reviewer 2: ```suggestion
description: Check your Apache Airflow DAGs for errors before you run it locally or deploy to Astro.
```

Let's try to have the word Airflow in as many of these as possible

---

> One of the Astro CLI's main features is the ability to run Astro projects in a local Airflow environment. It additionally includes commands that you can use to test and debug DAGs both inside and outside of a locally running Airflow environment. Use the following document to learn more about how you can test locally with the Astro CLI before deploying your code changes to a production environment.

Reviewer 2: I would expand this intro more with a bullet list of the kinds of tests the CLI supports / thus what is covered in this doc. I think we're underselling what the Astro CLI can do

Reviewer: @paola - would you be okay with us addressing this change as part of the follow-up work to separate it into Test DAGs + Test projects?

> One of the Astro CLI's main features is the ability to run Astro projects in a local Airflow environment. It additionally includes commands that you can use to test and debug DAGs both inside and outside of a locally running Airflow environment. Use the following document to learn more about how you can test locally with the Astro CLI before deploying your code changes to a production environment.

Author: ```suggestion
One of the Astro CLI's main features is the ability to run Astro projects in a local Airflow environment. It additionally includes commands that you can use to test and debug DAGs both inside and outside of a locally running Airflow environment.

Use the following document to learn more about how you can test locally with the Astro CLI before deploying your code changes to a production environment.
```

To run a DAG located within your local `/dags` directory run:

```sh
astro run <dag-id>
```

You can only run one DAG at a time. All the tasks in your DAG run sequentially. Any errors produced by your code while parsing or running your DAG appear in the command line. For more information about this command, see the [CLI command reference](cli/astro-run.md).

## Unit test DAGs 

You can run DAG unit tests with the following Astro CLI commands to quickly test code:

- `astro dev parse`
- `astro dev pytest`

These commands don't require a running Airflow environment, meaning you can test DAGs without deploying to Astro or starting up Docker. 

### Parse DAGs

To quickly parse your DAGs, run:

```sh
astro dev parse
```

> 

Author: Yep!

### Run tests with pytest

To run unit tests on your Astro project, run:

```sh
astro dev pytest
```

This command runs all tests in your project's `tests` directory with [pytest](https://docs.pytest.org/en/7.0.x/index.html#), a testing framework for Python. With pytest, you can test custom Python code and operators locally without having to start a local Airflow environment.

By default, the `tests` directory in your Astro project includes a default DAG integrity test called `test_dag_integrity.py`. This test checks that:

- All Airflow tasks have required arguments.
- DAG IDs are unique across the Astro project.
- DAGs have no cycles.
- There are no general import or syntax errors.

`astro dev pytest` runs this default test alongside any other custom tests that you add to the `tests` directory. For more information about this command, see the [CLI command reference](cli/astro-dev-pytest.md).

> ## Test before upgrading your Astro project

Reviewer 2: Should this title be more direct? `Test an Airflow version upgrade`

Author: Sure but I think we need to put Runtime here, it's the only term we refer to in the topic

> ## Test before upgrading your Astro project

Author: ```suggestion
## Test before an Astro Runtime upgrade
```

---

> sidebar_label: 'Test your project'

Author: ```suggestion
sidebar_label: 'Test your DAGs'
```

title: 'Test your Astro project'
id: test-your-astro-project-locally
description: Check your Airflow DAGs for errors before you run it locally or deploy to Astro.
---

One of the Astro CLI's main features is the ability to run Astro projects in a local Airflow environment. It additionally includes commands that you can use to test and debug DAGs both inside and outside of a locally running Airflow environment.

Use the following document to learn more about how you can test locally with the Astro CLI before deploying your code changes to a production environment.

## Run a DAG with `astro run`

Use the `astro run` command to run a DAG from the command line. When you run the command, the CLI compiles your DAG and runs it in a single Airflow worker container based on your Astro project configurations, including your `Dockerfile`, DAG utility files, Python requirements, and environment variables. You can review task logs and task status in your terminal without opening the Airflow UI. Running DAGs without a scheduler or webserver can help reduce the time required to develop and test data pipelines.

> To run a DAG located within your local `/dags` directory run:

Author: ```suggestion
To run a DAG located within your local `/dags` directory, run:
```

```bash
astro run <dag-id>
```

You can only run one DAG at a time. All the tasks in your DAG run sequentially. Any errors produced by your code while parsing or running your DAG appear in the command line. For more information about this command, see the [CLI command reference](cli/astro-run.md).

## Unit test DAGs 

You can run DAG unit tests with the following Astro CLI commands to quickly test code:

- `astro dev parse`
- `astro dev pytest`

These commands don't require a running Airflow environment, meaning you can test DAGs without deploying to Astro or running Airflow locally. However, these commands do require Docker.

### Parse DAGs

To quickly parse your DAGs, run:

```sh
astro dev parse
```

This command parses your DAGs to ensure that they don't contain any basic syntax or import errors and that they can successfully render in the Airflow UI.

> The command `astro dev parse` is a more convenient but less customizable version of `astro dev pytest`. If you don't have any specific test files that you want to run on your DAGs, Astronomer recommends using `astro dev parse` as your primary testing tool. For more information about this command, see the [CLI command reference](cli/astro-dev-parse.md).

Author: ```suggestion
`astro dev parse` is a more convenient but less customizable version of `astro dev pytest`. If you don't have any specific test files that you want to run on your DAGs, Astronomer recommends using `astro dev parse` as your primary testing tool. For more information about this command, see the [CLI command reference](cli/astro-dev-parse.md).
```
